apps/meeting/src/providers/VideoGridProvider/state.tsx

```diff
@@ -0,0 +1,510 @@
+    if (type === 'activeSpeaker') {
+      videoPreferences.add(
+        new VideoPreference(
+          attendeeId,
+          1,
+          mode === VideoGridMode.FeaturedView
+            ? TargetDisplaySize.High
+            : targetDisplaySize
+        )
+      );
+    }
+
+    // Set the common tiles to low priority
+    if (type === 'common') {
+      videoPreferences.add(
+        new VideoPreference(attendeeId, 2, targetDisplaySize)
+      );
+    }
+  }
+  priorityBasedPolicy.chooseRemoteVideoSources(videoPreferences.build());
+};
+
+export function reducer(state: State, { type, payload }: Action): State {
+  const { gridState, attendeeStates, videoSourceState } = state;
+
+  switch (type) {
+    case VideoGridAction.ResetVideoGridState: {
+      return initialState;
+    }
+    case VideoGridAction.UpdateAttendeeStates: {
+      const { roster } = payload;
+
+      // Remove attendee that left the meeting
+      for (const attendeeId of Object.keys(attendeeStates)) {
+        if (!isContentShare(attendeeId) && !(attendeeId in roster)) {
+          delete attendeeStates[attendeeId];
+        }
+      }
+
+      // Add attendee that joined the meeting
+      for (const attendeeId of Object.keys(roster)) {
+        const name = roster[attendeeId]?.name || '';
+
+        if (attendeeId in attendeeStates) {
+          attendeeStates[attendeeId].name = name;
+        } else {
+          attendeeStates[attendeeId] = {
+            attendeeId,
+            name,
+            videoEnabled: false,
+            bandwidthConstrained: false,
+          } as AttendeeState;
+        }
+      }
+
+      // Ensure the state of `videoEnabled` in the racing condition of UpdateAttendeeStates and UpdateVideoSources
+      // To do: Merge all the update actions to one
+      for (const attendeeId of videoSourceState.cameraSources) {
+        if (attendeeId in attendeeStates) {
+          attendeeStates[attendeeId].videoEnabled = true;
+        }
+      }
+
+      return {
+        ...state,
+        attendeeStates,
+      };
+    }
+    case VideoGridAction.UpdateVideoSources: {
+      const { videoSources } = payload as { videoSources: VideoSource[] };
+      const cameraSources: string[] = [];
+      const videoSourceIds = videoSources.map(
+        videoSource => videoSource.attendee.attendeeId
+      );
+
+      // Reset the `videoEnabled` of all attendeeStates
+      for (const attendee of Object.values(attendeeStates)) {
+        attendee.videoEnabled = false;
+      }
+
+      // Remove content share from attendeeStates,
+      // if content share is not in video sources
+      for (const attendeeId of Object.keys(attendeeStates)) {
+        if (
+          isContentShare(attendeeId) &&
+          !videoSourceIds.includes(attendeeId)
+        ) {
+          delete attendeeStates[attendeeId];
+        }
+      }
+
+      // Update the `videoEnabled` according to video sources
+      for (const attendeeId of videoSourceIds) {
```

Comment:
The first loop where yyou map to get `videoSourceIds` can be avoided right? You can just use:
```js
for (const videoSource of videoSources) {
const attendeeId = videoSource.attendee.attendeeId;
..rest same.
}

Reply:
We need to check whether content share id is in the `videoSourceIds`. So have to keep it, but to optimize the performance, I update the `videoSourceIds` from array to set `videoSourceIdSet`.